src/host/main.cpp

```diff
@@ -505,236 +505,252 @@ int main(int argc, char** argv)
-  files::dump(node_cert, node_cert_file);
-  if (*start || *recover)
-  {
-    files::dump(network_cert, network_cert_file);
-    files::dump(network_enc_pubk, network_enc_pubk_file);
-  }
+    if (start_type == StartType::Unknown)
+    {
+      LOG_FATAL_FMT("Start command should be start|join|recover. Exiting.");
+    }
 
-  auto enclave_thread_start = [&]() {
-#ifndef VIRTUAL_ENCLAVE
-    try
-#endif
+    enclave.create_node(
+      enclave_config,
+      ccf_config,
+      node_cert,
+      network_cert,
+      network_enc_pubk,
+      start_type,
+      consensus,
+      num_worker_threads,
+      time_updater->behaviour.get_value());
+
+    LOG_INFO_FMT("Created new node");
+
+    // Write the node and network certs to disk.
+    files::dump(node_cert, node_cert_file);
+    if (*start || *recover)
     {
-      enclave.run();
+      files::dump(network_cert, network_cert_file);
+      files::dump(network_enc_pubk, network_enc_pubk_file);
     }
+
+    auto enclave_thread_start = [&]() {
 #ifndef VIRTUAL_ENCLAVE
-    catch (const std::exception& e)
+      try
+#endif
+      {
+        enclave.run();
+      }
+#ifndef VIRTUAL_ENCLAVE
+      catch (const std::exception& e)
+      {
+        LOG_FAIL_FMT("Exception in enclave::run: {}", e.what());
+
+        // This exception should be rethrown, probably aborting the process, but
+        // we sleep briefly to allow more outbound messages to be processed. If
+        // the enclave sent logging messages, it is useful to read and print
+        // them before dying.
+        std::this_thread::sleep_for(1s);
+        throw;
+      }
+#endif
+    };
+
+    // Start threads which will ECall and process messages inside the enclave
+    std::vector<std::thread> threads;
+    for (uint32_t i = 0; i < (num_worker_threads + 1); ++i)
     {
-      LOG_FAIL_FMT("Exception in enclave::run: {}", e.what());
-
-      // This exception should be rethrown, probably aborting the process, but
-      // we sleep briefly to allow more outbound messages to be processed. If
-      // the enclave sent logging messages, it is useful to read and print
-      // them before dying.
-      std::this_thread::sleep_for(1s);
-      throw;
+      threads.emplace_back(std::thread(enclave_thread_start));
     }
-#endif
-  };
 
-  // Start threads which will ECall and process messages inside the enclave
-  std::vector<std::thread> threads;
-  for (uint32_t i = 0; i < (num_worker_threads + 1); ++i)
-  {
-    threads.emplace_back(std::thread(enclave_thread_start));
+    uv_run(uv_default_loop(), UV_RUN_DEFAULT);
+    for (auto& t : threads)
+    {
+      t.join();
+    }
   }
 
-  uv_run(uv_default_loop(), UV_RUN_DEFAULT);
-  for (auto& t : threads)
+  // Continue running the loop long enough for the on_close
+  // callbacks to be despatched, so as to avoid memory being
+  // leaked by handles. Capped out of abundance of caution.
+  size_t close_iterations = 100;
+  while (close_iterations > 0 && uv_loop_alive(uv_default_loop()))
```

Comment:
Is there an issue here? if close_iterations == 0 && uv_loop_alive(uv_default_loop()) == true then on the next iterations close iterations will be size_t::max

Reply:
If close_iterations == 0 we do not enter the loop, because !(close_iterations > 0), and the value of loop_alive is immaterial because it's an && ?